src/librustc/ty/context.rs

```diff
@@ -1408,6 +1418,22 @@ impl<'a, 'gcx, 'tcx> TyCtxt<'a, 'gcx, 'tcx> {
         self._intern_existential_predicates(eps)
     }
 
+    pub fn intern_predicates(self, preds: &[Predicate<'tcx>])
+        -> &'tcx Slice<Predicate<'tcx>> {
+        // FIXME consider asking the input slice to be sorted to avoid
+        // re-interning permutations, in which case that would be asserted
+        // here.
+        if preds.len() == 0 {
+            // The macro-generated method below asserts we don't intern an empty slice.
+            // FIXME: intern_type_list also dishes out new empty Slices on every invocation,
+            // so is this the right thing to do here as well? We should be able to cache things
+            // in this case eventually.
```

Comment:
Generating a "fresh slice" is basically free, so caching it makes no sense. You can remove this FIXME.